options/rtld/generic/linker.cpp

```diff
@@ -254,6 +254,39 @@ frg::expected<LinkerError, SharedObject *> ObjectRepository::requestObjectWithNa
 		mlibc::infoLogger() << "rtld: no rpath set for object" << frg::endlog;
 	}
 
+	if(fd && !chosenPath.empty()) {
+		if (createScope) {
+			__ensure(localScope == nullptr);
+
+			// TODO: Free this when the scope is no longer needed.
+			localScope = frg::construct<Scope>(getAllocator());
+		}
+
+		__ensure(localScope != nullptr);
+
+		auto object = frg::construct<SharedObject>(getAllocator(),
+			name.data(), std::move(chosenPath), false, localScope, rts);
+
+		auto result = _fetchFromFile(object, fd);
+		closeOrDie(fd);
+		if(!result) {
+			frg::destruct(getAllocator(), object);
+			if (createScope) {
+				frg::destruct(getAllocator(), localScope);
+				localScope = nullptr;
+			}
+			return result.error();
+		}
+
+		_parseDynamic(object);
+		_parseVerdef(object);
+		_addLoadedObject(object);
+
+		return object;
+	}
```

Comment:
This logic should not be duplicated. This should be moved into an own function that is called both on the RPATH code path and on the library path code path. Also, error handling can potentially be improve with scope_exit.